### PR TITLE
Honor “lang” attribute in no namespace only on HTML and SVG elements

### PR DESCRIPTION
**Pull request by sideshowbarker**
#### 6599021230e5957efed742285983012bd5b5703f
<pre>
Honor “lang” attribute in no namespace only on HTML and SVG elements
<a href="https://bugs.webkit.org/show_bug.cgi?id=263655">https://bugs.webkit.org/show_bug.cgi?id=263655</a>

Reviewed by Ryosuke Niwa.

<a href="https://github.com/whatwg/html/pull/9882/files">https://github.com/whatwg/html/pull/9882/files</a> added the phrase “or an
element in the SVG namespace” in the following HTML spec text:

&gt; To determine the language of a node, user agents must use the first
appropriate step in the following list:
&gt; …
&gt; - If the node is an HTML element or an element in the SVG namespace,
    and it has a lang in no namespace attribute set: Use the value of
    that attribute.

And that updated spec text is now part of the HTML standard, at
<a href="https://html.spec.whatwg.org/#the-lang-and-xml">https://html.spec.whatwg.org/#the-lang-and-xml</a>:lang-attributes:svg-namespace

Gecko has already had that behavior implemented for quite a while, and
Blink was recently updated to implement it as well.

So this change fully aligns WebKit with that updated spec text, and with
the current behavior in Gecko and Blink. However, this change does that
by updating the code to *restrict* the “lang” attribute in no namespace
to *only* being honored on HTML and SVG elements.

Otherwise, without this code change, WebKit honors the “lang” attribute
in no namespace regardless of what element it’s specified on (that is,
not just for HTML elements and SVG elements) — which breaks conformance
with the spec, and breaks interoperability with Gecko and Blink.

* LayoutTests/imported/w3c/web-platform-tests/html/dom/elements/global-attributes/lang-attribute.window-expected.txt:
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::attributeChanged):

Canonical link: <a href="https://commits.webkit.org/273726@main">https://commits.webkit.org/273726@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/67f34695bd05992f99053e13c1ee086c5921cf2e

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/36409 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/15359 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/38627 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/39114 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/32706 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/37639 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/17811 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/12392 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/31352 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/36970 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/12959 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/32261 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/11356 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/11384 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/32494 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/40359 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/33026 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/32843 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/37322 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/11601 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/9475 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/35422 "Passed tests") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/13309 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/8264 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/12051 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/12468 "Built successfully") | | |

### CODE REVIEW

**Review by annevk — COMMENTED** [review]
Non-reviewer caveats apply: I think we should instead update `hasLangAttr()` to include these checks. Because now other things building on top of `hasLangAttr()` become incorrect of which you caught some, but not in the most elegant way (it seems more expensive to turn has checks into checks on returned strings) and it's also not clear to me you caught all as `hasLangAttr()` is still around and `hasLanguageAttribute()` is too.

Hope that helps!

Aside: looking at the code it seems we don't support shadow tree inheritance for language either. Maybe a fun follow-up issue?

**Comment by sideshowbarker** [comment]
> I think we should instead update `hasLangAttr()` to include these checks.

Indeed — thanks, that’s clearly the right place for the change. Made it so.

> Because now other things building on top of `hasLangAttr()` become incorrect of which you caught some, but not in the most elegant way (it seems more expensive to turn has checks into checks on returned strings) and it's also not clear to me you caught all as `hasLangAttr()` is still around and `hasLanguageAttribute()` is too.

Yup — changing `hasLangAttr()` instead solves that problem.

> Aside: looking at the code it seems we don't support shadow tree inheritance for language either. Maybe a fun follow-up issue?

Yeah, I had noticed the older existing test at https://github.com/web-platform-tests/wpt/blob/master/html/dom/elements/global-attributes/lang-attribute-shadow.window.js but I hadn’t made time yet to figure out what part of the spec states the corresponding requirement. Once I find the actual spec statement of the requirement, I’ll open another PR with an implementation.

**Review by annevk — COMMENTED** [review]
The logic of this seems sound. Thanks!

Now you need an actual WebKit reviewer for the remainder. Maybe @cdumez or @rniwa?

**Comment by sideshowbarker** [comment]
> > Aside: looking at the code it seems we don't support shadow tree inheritance for language either. Maybe a fun follow-up issue?
> 
> Yeah, I had noticed the older existing test at [web-platform-tests/wpt@`master`/html/dom/elements/global-attributes/lang-attribute-shadow.window.js](https://github.com/web-platform-tests/wpt/blob/master/html/dom/elements/global-attributes/lang-attribute-shadow.window.js?rgh-link-date=2023-10-31T22%3A07%3A16Z) but I hadn’t made time yet to figure out what part of the spec states the corresponding requirement. Once I find the actual spec statement of the requirement, I’ll open another PR with an implementation.

Only after writing that did I realize that requirement is in fact the very next step in the spec algorithm…

Anyway, https://wpt.fyi/results/html/dom/elements/global-attributes/lang-attribute-shadow.window.html shows that test already passing in TP181. And manual test confirms that it does actually already pass both in TP181 and Safari 17.2

So it seems like there’s nothing more needs for that — as far as implementation goes at least?

I do notice that the `lang-attribute-shadow.window.html` test hasn’t yet been imported though. I don’t know what the project protocol is in that case — but if it seems like we should import that test, I can open a PR for importing it.

**Comment by sideshowbarker** [comment]
> Anyway, [wpt.fyi/results/html/dom/elements/global-attributes/lang-attribute-shadow.window.html](https://wpt.fyi/results/html/dom/elements/global-attributes/lang-attribute-shadow.window.html) shows that test already passing in TP181. And manual test confirms that it does actually already pass both in TP181 and Safari 17.2

For the record here: It turns out WebKit is passing now because the recent test change (and the spec change it was based on) were reverted a few days ago. See https://bugs.webkit.org/show_bug.cgi?id=263655#c5

So I guess this is good example of a case where it paid off to wait for a while before implementing a (very recent) spec change :p

Though that said, I’ll confess now that yesterday I had actually written (locally) a WebKit patch that implements the current spec text (which doesn’t include the https://github.com/whatwg/html/pull/9880 revert, which hasn’t been merged yet), and I had been surprised yesterday when I tried it with the current http://wpt.live/html/dom/elements/global-attributes/lang-attribute-shadow.window.html test and found that it was failing. But I see now that’s because the https://github.com/web-platform-tests/wpt/pull/42811/files#diff-3e68420ba9732782c4a09ec7119cd7556e9d1aedef5f2bc98f8d7ef92322a4ab WPT update/revert had already been merged (while the spec change hasn’t yet been)

So it did cost me some time yesterday to write up the patch — but at least it didn’t cost anybody else any time to review it (since I hadn’t yet gotten around to opening a PR for it).

Anyway, that’s actually all completely orthogonal to the change in this PR branch, so I’ll be quiet about it now…

**Review by annevk — COMMENTED** [review]
Not actually having looked at benchmark numbers I suspect it would be better if we instead changed when we called `setHasLangAttr()`. In particular with selectors we'd do more lookups and so having that be a single bit check seems best.

And unfortunately this also needs rebasing due to Ahmad's recent work on importing that directory.

**Comment by sideshowbarker** [comment]
> Not actually having looked at benchmark numbers I suspect it would be better if we instead changed when we called `setHasLangAttr()`. In particular with selectors we'd do more lookups and so having that be a single bit check seems best.

Thanks, makes sense — made it so

**Comment by annevk** [comment]
This looks good to me, but I asked a more experienced DOM reviewer to do a final review.

File: Source/WebCore/dom/Element.cpp [hunk 1]
```diff
@@ -2141,7 +2141,7 @@ void Element::attributeChanged(const QualifiedName& name, const AtomString& oldV
     case AttributeNames::XML::langAttr:
     case AttributeNames::langAttr: {
         if (name == HTMLNames::langAttr)
-            setHasLangAttr(!newValue.isNull());
+            setHasLangAttr(!newValue.isNull() && ((isHTMLElement() || isSVGElement())));
```
Comment (rniwa) [inline Source/WebCore/dom/Element.cpp]: Nit: You have an extra parenthesis here. Just do: `!newValue.isNull() && (isHTMLElement() || isSVGElement())`

Comment (sideshowbarker) [reply Source/WebCore/dom/Element.cpp]: Thanks for catching that — fixed now

**Review by rniwa — APPROVED** [review]

**Pull request merged**